File: exercises/python/lab09/msg_reader.py
"""Lab 09 - Message Queue IPC (Reader) using queue + file"""
import pickle
import os

QUEUE_FILE = "/tmp/os_lab09_msgqueue"


def main():
    if not os.path.exists(QUEUE_FILE):
        print("No message queue found. Run msg_writer.py first.")
        return

    with open(QUEUE_FILE, "rb") as f:
        messages = pickle.load(f)

    if not messages:
        print("Message queue is empty.")
        return

    # Receive the first type-1 message
    msg = None
    for i, m in enumerate(messages):
        if m["type"] == 1:
            msg = messages.pop(i)
            break

    if msg is None:
        print("No type-1 message found.")
        return

    # Save remaining messages back
    if messages:
        with open(QUEUE_FILE, "wb") as f:
            pickle.dump(messages, f)
    else:
        os.remove(QUEUE_FILE)
        print("Message queue destroyed.")

    print(f'Message received: "{msg["text"]}"')


if __name__ == "__main__":
    main()

File: exercises/cpp/lab09/msg_reader.cpp
// Lab 09 - Message Queue IPC (Reader)
#include <stdio.h>
#include <sys/ipc.h>
#include <sys/msg.h>

#define MAX 100

struct mesg_buffer {
    long mesg_type;
    char mesg_text[MAX];
};

int main() {
    key_t key  = ftok("progfile", 65);
    if (key == -1) { perror("ftok (run msg_writer first)"); return 1; }
    int   msgid = msgget(key, 0666 | IPC_CREAT);
    if (msgid == -1) { perror("msgget"); return 1; }

    struct mesg_buffer message;
    if (msgrcv(msgid, &message, sizeof(message), 1, 0) == -1) {
        perror("msgrcv"); return 1;
    }

    printf("Message received: \"%s\"\n", message.mesg_text);

    // Destroy the message queue
    msgctl(msgid, IPC_RMID, NULL);
    printf("Message queue destroyed.\n");
    return 0;
}
